query/src/main/scala/filodb/query/exec/PeriodicSamplesMapper.scala

```diff
@@ -99,68 +99,64 @@ final case class PeriodicSamplesMapper(start: Long,
  * However, note that sliding window iterators have to do twice as much work, adding and removing, so it would
  * only be worth it if the overlap is more than 50%.
  */
-//scalastyle:off null
 class ChunkedWindowIterator(rv: RawDataRangeVector,
                             start: Long,
                             step: Long,
                             end: Long,
                             window: Long,
                             rangeFunction: ChunkedRangeFunction,
                             queryConfig: QueryConfig) extends Iterator[TransientRow] with StrictLogging {
-  private var windowIt: WindowedChunkIterator = null
+  // Lazily open the iterator and obtain the lock. This allows one thread to create the
+  // iterator, but the lock is owned by the thread actually performing the iteration.
+  private lazy val windowIt = {
+    val it = new WindowedChunkIterator(rv.chunkInfos(start - window, end), start, step, end, window)
+    // Need to hold the shared lock explicitly, because the window iterator needs to
+    // pre-fetch chunks to determine the window. This pre-fetching can force the internal
+    // iterator to close, which would release the lock too soon.
+    it.lock()
+    it
+  }
 
   private val sampleToEmit = new TransientRow()
 
-  override def hasNext: Boolean = openWindowIt().hasMoreWindows
+  override def hasNext: Boolean = windowIt.hasMoreWindows
 
   override def next: TransientRow = {
     rangeFunction.reset()
     // TODO: detect if rangeFunction needs items completely sorted.  For example, it is possible
     // to do rate if only each chunk is sorted.  Also check for counter correction here
 
-    val windowIt = openWindowIt()
+    // Lazy variables have an extra lookup cost, due to a volatile bitmap field generated by
+    // the compiler. Copy to a local variable to reduce some overhead.
+    val wit = windowIt
```

Comment:
Hmmm.  When you declare a val in a class, it is initialized in the constructor, thus rendering the original lazy val code useless.  Lazy Vals are fairly inexpensive, but if we are OK with initializing in the constructor, why not just use a val, then we don't need the extra 2 fields for the lazy val?  and I don't see why we wouldn't be OK with the initialization at constructor time (other than minor edge case where the windowed iterator is not used at all....)

Reply:
I disassembled the code and verified that the internal iterator is not constructed until it's first used. The comment on the lazy val explains why it's done this way. The thread creating ChunkedWindowIterator isn't necessarily the thread which actually uses it.